### DISCUSSION TITLE
Experiment and browser compatibility?

### DISCUSSION THREAD

**Discussion by jubejubess** [jspsych/jsPsych#797] [Q&A]
Hello!

We currently are running a few experiments created through jsPsych and hosted on JATOS. We noticed that all of our experiments seem to work decently well (so far) using Firefox and Google Chrome, but they all fail to load using any other browser (safari, microsoft edge, explorer). For reference, one experiment preloads about 150 audio stimuli, another experiment only preloads 9 audio stimuli while the other experiment preloads about 8 video files.  

Just wondering if anyone knew what could be an underlying issue? 

This is the error code that appears when the experiment fails to load in the other browsers:

``` TypeError: onLoadCallback is not a function. (In 'onLoadCallback()', 'onLoadCallback' is an instance of Object) ```

Thank you!

**Comment by kristian-lange-tw** [comment]
Hi!
Looks like some problem with jatos.js initialization. Can you please post here the code where you do your `jatos.onLoad` and `jsPsych.init`. And which JATOS version do you use locally and on your server?
Best,
Kristian

**Reply by jubejubess** [reply]
Hi!, this is code for two of our experiments. We use version 3.3.6 for both versions! 

```
jatos.onLoad(
		jatos.addJatosIds(resultData),

		jsPsych.init({
			timeline: timeline,

			exclusions: {
          	  min_width: 800,
          	  min_height: 600
        },
        on_interaction_data_update: function (data) {
            if (data.event == "blur") {
                console.log(data);
                blur_count++;
                if (blur_count > 10) {
                    likely_invalid = true;
                }
            };
        },

			use_webaudio: false,
			preload_audio: audio,
			max_load_time: 12000,
			max_preload_attempts: 10,			

			//to display data and add subject number:
			on_finish: function() {
				var studyID = jatos.studyResultID;
				 if(likely_invalid){
                	studyID += ' - invalid result'
            }
				jsPsych.data.addProperties({subject : studyID});
				// jsPsych.data.displayData("CSV");
				//to submit the results to JATOS...
        	    var all_data = jsPsych.data.get();
        	    var results = all_data.ignore('internal_node_id').ignore('time_elapsed');
            	jatos.submitResultData(results.csv(), jatos.startNextComponent);
			}
		})
	);
```


```
jatos.onLoad(function() {
var condition = jatos.urlQueryParameters.condition;   // accesses condition number from the URL
var studyID = jatos.studyResultId;  // creates a study ID 
jsPsych.data.addProperties({subject : studyID});     // adds study ID to results data
jsPsych.data.addProperties({condition : condition});   // adds condition to results data

jsPsych.init({
        timeline: timeline, // plays timeline "trial"
        use_webaudio: false, // allows audio to play in browser
        show_preload_progress_bar: false, // hide preload progress bar,
        preload_video: ["video/VbA0.mp4", "video/VbA4.mp4", "video/VbA5.mp4", "video/VbA6.mp4", "video/VdA0.mp4", "video/VdA4.mp4", "video/VdA5.mp4", "video/VdA6.mp4"],
        preload_audio: ["audio/bd04.wav", "audio/bd04_minus.wav", "audio/bd04_plus.wav", "audio/bd05.wav", "audio/bd05_minus.wav", "audio/bd05_plus.wav", "audio/bd06.wav", "audio/bd06_plus.wav", "audio/bd06_minus.wav"],
                    exclusions: {
                        min_width: 800,
                        min_height: 600
                    },
        on_finish: function (data) {
          var resultJson = jsPsych.data.get().ignore('internal_node_id').ignore('time_elapsed'); // retrieves all data
          jatos.submitResultData(resultJson.csv(), jatos.startNextComponent); 
        },
        default_iti: 10
      });
    });
```

**Comment by kristian-lange** [comment]
In your first experiment you do not pass a function to `jatos.onLoad`.
Instead of
`jatos.onLoad(`
you should do 
`jatos.onLoad(function () { `
as you do in your second experiment. 

It looks like Chrome and FF were just not very strict in JavaScript syntax.

**Reply by jubejubess** [reply]
Hello!

Thank you! I tried changing it in the first experiment, but both experiments still don't seem to load for safari.
In the past I've sometimes encountered this error in the console: "Failed to load resource: the server responded with a status of 404 (Not Found)." The requested URL /favicon.ico was not found on this server.
While, other times the console doesn't report any errors.

**Comment by kristian-lange** [comment]
This error "Failed to load resource: ... favicon.ico was not found on this server." is nothing to worry about. It just states that the browser couldn't find a favicon (those little logos in your browser's tabs). But the page will still work fine. 

Apart from the favicon error msg do you get any other ones?

**Reply by jubejubess** [reply]
Ah okay! Other than that, I haven’t seen any other errors that occur in the console! :/ Would it be an issue related to how safari deals with media playback?

**Reply by kristian-lange** [reply]
To be honest I don't know how Safari deals with media playback. I do not use MacOS (anymore). Is there something worrisome?

**Reply by jubejubess** [reply]
I'm not sure -- I asked this in another thread, but I tested a smaller scale experiment and it worked in most browsers, but the audio playback was not in sync for safari and explorer which made me initially think that it couldn't handle media playback at a larger scale as well as other browsers!

**Reply by parkerrobbins** [reply]
Hi, @jubejubess! Correct me if I'm wrong, but I believe #809 is the related thread you are referring to.

**Reply by jubejubess** [reply]
Hi! Yes, that's the other thread!
